server/server_test.go

```diff
@@ -0,0 +1,94 @@
+package server_test
+
+import (
+	"context"
+	"database/sql"
+	"net"
+	"testing"
+
+	vsql "github.com/dolthub/vitess/go/mysql"
+	"github.com/go-sql-driver/mysql"
+
+	sqle "github.com/dolthub/go-mysql-server"
+	"github.com/dolthub/go-mysql-server/memory"
+	"github.com/dolthub/go-mysql-server/server"
+	gsql "github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/mysql_db"
+	"google.golang.org/grpc/test/bufconn"
+)
+
+// TestSeverCustomListener verifies a caller can provide their own net.Conn implementation for the server to use
+func TestSeverCustomListener(t *testing.T) {
+	dbName := "mydb"
+	// create a net.Conn thats based on a golang buffer
+	buffer := 1024
+	listener := bufconn.Listen(buffer)
+
+	// create the memory database
+	memdb := memory.NewDatabase(dbName)
+	pro := memory.NewDBProvider(memdb)
+	engine := sqle.NewDefault(pro)
+
+	// server config with custom listener
+	cfg := server.Config{Listener: listener}
+	// since we're using a memory db, we can't rely on server.DefaultSessionBuilder as it causes panics, so explicitly build a memorySessionBuilder
+	sessionBuilder := func(ctx context.Context, c *vsql.Conn, addr string) (gsql.Session, error) {
+		host := ""
+		user := ""
+		mysqlConnectionUser, ok := c.UserData.(mysql_db.MysqlConnectionUser)
+		if ok {
+			host = mysqlConnectionUser.Host
+			user = mysqlConnectionUser.User
+		}
+		client := gsql.Client{Address: host, User: user, Capabilities: c.Capabilities}
+		return memory.NewSession(gsql.NewBaseSessionWithClientServer(addr, client, c.ConnectionID), pro), nil
+	}
+	s, err := server.NewServer(cfg, engine, sessionBuilder, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	go func() {
+		if err = s.Start(); err != nil {
```

Comment:
You've got a race here. Need to wait for the server to become available, or else add retry logic to Ping below

Reply:
Add a simple retry loop to the ping- let me know if you have further opinions on how that should behave 👍